code/mob/living/critter/ai/sawflyai.dm

```diff
@@ -0,0 +1,175 @@
+//For the sawfly critter itself, check mob/living/critter/sawfly.dm
+//For misc things, check sawflymisc.dm in code/obj.sawflymisc
+
+/datum/aiHolder/sawfly
+
+	New()
+		..()
+		default_task = get_instance(/datum/aiTask/prioritizer/sawfly, list(src))
+
+/datum/aiTask/prioritizer/sawfly
+	name = "base thinking (should never see this)"
+
+/datum/aiTask/prioritizer/sawfly/New()
+	..()
+	// populate the list of tasks
+	transition_tasks += holder.get_instance(/datum/aiTask/timed/wander, list(holder, src))
+	transition_tasks += holder.get_instance(/datum/aiTask/timed/targeted/sawfly_attack, list(holder, src))
+	transition_tasks += holder.get_instance(/datum/aiTask/sequence/goalbased/sawfly_chase_n_stab, list(holder, src))
+
+/datum/aiTask/prioritizer/sawfly/on_tick()
+	if(isdead(holder.owner))
+		holder.enabled = FALSE
+		walk(holder.owner, 0)
+
+/datum/aiTask/prioritizer/sawfly/on_reset()
+	..()
+	walk(holder.owner, 0)
+
+// Custom behaviour starts here
+
+// Attack code - looks for adjacent targets, tries to stab them
+/datum/aiTask/timed/targeted/sawfly_attack
+	var/found_path = null
+	name = "attacking"
+	minimum_task_ticks = 10
+	maximum_task_ticks = 25
+	var/weight = 10
+	target_range = 1
+	var/list/dummy_params = list("icon-x" = 16, "icon-y" = 16)
+
+/datum/aiTask/timed/targeted/sawfly_attack/evaluate()
+	return weight * score_target(get_best_target(get_targets()))
+
+/datum/aiTask/timed/targeted/sawfly_attack/on_tick()
+	var/mob/living/critter/robotic/sawfly/owncritter = holder.owner
+	if(prob(5)) owncritter.communalbeep()
+	walk(owncritter, 0)
+	if(!holder.target)
+		holder.target = get_best_target(get_targets())
+	if(holder.target)
+		var/atom/T = holder.target
+		// if target is dead
+		// fetch a new one if we can
+		if(isliving(T))
+			var/mob/living/M = T
+			if(M.health < -50)
+				holder.target = get_best_target(get_targets())
+			if(istype(T, /mob/living/critter/robotic/sawfly))
+				holder.target = get_best_target(get_targets())
+		if(!holder.target) //we lost target, re-evaluate tasks
+			holder.interrupt()
+			return
+
+		var/dist = get_dist(owncritter, holder.target)
+		if(dist > target_range)
+			if(!src.found_path)
+				src.found_path = get_path_to(holder.owner, holder.target, 18, 0)
+			if(src.found_path)
+				holder.move_to_with_path(holder.target, src.found_path, 1)
+				if(prob(20)) walk_rand(src,4)
```

Comment:
This is gonna do some weird stuff with the pathing code. `move_to_with_path` gets a list of tiles to step on, each one adjacent to the next. It has the assumption that if you are about to step on tile 2, it's because you're already standing on tile 1. `walk_rand` will break break that assumption, resulting in the sawfly trying step from some random tile on to tile 2, which will *probably* be fine if it happens only once, but might get fucky if it keeps happening. 
Not saying this necessarily needs to be removed, but it should be thoroughly tested to ensure you don't get weird behaviour like the drone suddenly walking through a wall or rocketing across a room.

Reply:
Tested this a ton and I'm proud to announce they're perfectly stable!